### ISSUE TITLE
don't perform verification when there are 0 proof obligations

### ISSUE THREAD

**Issue by 0xabu** [boogie-org/boogie#43]
I have a large project that contains lots of very simple functions. I am consequently spending lots of my time waiting for things such as the following:

```
Verifying CheckWellformed$$_module.__default.NR__L1PTES ...
  [0.647 s, 0 proof obligations]  verified

Verifying CheckWellformed$$_module.__default.NR__L2PTES ...
  [0.717 s, 0 proof obligations]  verified

Verifying CheckWellformed$$_module.__default.reveal__smc__remove__premium ...
  [0.923 s, 0 proof obligations]  verified

Verifying CheckWellformed$$_module.__default.reveal__smc__mapSecure__premium ...
  [0.880 s, 0 proof obligations]  verified
```

It would be nice if Boogie could skip invoking the verifier if there are really no proof obligations.

**Comment by shazqadeer** [comment]
@0xabu : Sorry it has taken almost four years to respond to this issue.  

I first educated myself on what the number of proof obligations means.  Turns out it means the total number of assert statements in the desugared code of the  implementation being verified.  By desugared code, I mean the code obtained by converting all specifications including preconditions, postconditions, loop invariants, etc. into assert and assume statements.  0 proof obligations means that there were no assert statements in the desugared code, so nothing to verify.

I am surprised by the time taken for each implementation in the output snippet shown by you.  Running times between 0.5 and 1 s are really large even if the theorem prover is invoked.  In addition to the actual invocation of SMT solver, there is time being spent in converting each implementation into a logical formula (this is where the total number of assertions is estimated, for example).  So, it is possible that each of your implementations is really large but with nothing to verify.  If so, this would be an "edge" case whose optimization might be wiser to leave to the verification tool that targets Boogie.

In any case, to take more action, we would need more investigation from your end.  In case you have moved on and are no longer interested in further investigation, let us know so I can close this issue.

**Comment by 0xabu** [comment]
Hi Shaz,

the verification tool here was the Dafny backend to [Vale](https://github.com/project-everest/vale), and the sources to it were [Komodo](https://github.com/Microsoft/Komodo). Since Vale is primarily a mechanical translation of text into Dafny, and then Dafny does another mechanical translation to Boogie, I think both tools lacks sufficient knowledge of what they might (or in this case might not) be asking Boogie to verify.

That said, yes, I've moved on.

**Comment by RustanLeino** [comment]
Dafny does a check (which I think was triggered by this concern in Komodo a few years ago) to see if the Boogie `implementation` it's about to give to Boogie has 0 proof obligations (with the same definition as Shaz gives above). If the `implementation` has 0 proof obligations, Dafny omits (rather than emits) it.

It would be nicer to have this functionality in Boogie, but from Dafny's perspective, at least the issue has been addressed. The mystery of why these 0-proof-obligations implementations cost so much time was never resolved.

**Comment by shazqadeer** [comment]
I am not convinced that detecting the absence of asserts, although it could be implemented, is the solution to your problem.  I do not expect VC checking to be more than a few milliseconds for trivial VCs (time taken is usually a small fraction of a second for ordinary VCs).  The real problem to debug is why these trivial VCs are taking such a long time.  Is there any chance that those Komodo files demonstrating this problem could be resurrected?  If not, I will close this issue and simply keep an eye out for next time this particular phenomenon pops up.

**Comment by 0xabu** [comment]
I'm not sure whether this will help, but I dug up a Komodo enlistment, and I'm attaching the Boogie output generated by dafny's `/print:` option for the file in question:
[smc_handler.gen.bpl.gz](https://github.com/boogie-org/boogie/files/4983412/smc_handler.gen.bpl.gz)

**Comment by shazqadeer** [comment]
I ran the version of Boogie in master on the file you sent me. It ran and even verified 6 implementations before getting stuck. Amazing!  It would be nice to know the options used by Dafny for invoking Boogie.  Perhaps that would make the verification go faster.

The file is really large, more than 350KLOC.  It has only ten implementations.  So, each implementation is likely extremely large.  If this file is representative of Boogie code being generated in Komodo, a running time of 0.5-1 s even for an implementation that has zero proof obligations is not surprising.  I can believe that for such  large implementations, VC generation itself could take that much time.

VC generation in Boogie (although worst-case quadratic in the size of the implementation) is expected to be linear time.  Scanning the implementation for a proof obligation is best done once the implementation has been desugared but before the VC generation is performed.  So the most we can save is the cost of the VC generation.  I will try to measure the time taken for different phases of VC generation for each implementation in the file you sent me and report back.

**Comment by 0xabu** [comment]
The full set of flags we pass to Dafny are specified [here](https://github.com/microsoft/Komodo/blob/master/verified/subdir.mk#L3-L10), but I'm not sure how those translate to Boogie. If I had to hazard a guess, you're most likely getting stuck on non-linear arithmetic (we pass /noNLarith to Dafny). We also have a fairly high timeout, and of course this work was all done on a ~4 year old version of Boogie and Z3, so I wouldn't expect everything to verify with modern tools.

As for the size: yes, the combination of Vale and Dafny led to a lot of output. This file in particular is the top-level implementation file in the system (it includes almost all the rest of the implementation) so was probably the largest input to Boogie, and thus the worst culprit for related scaling issues.

**Comment by shazqadeer** [comment]
fixed by PR #389
